LevelDB Wrapper

aostrun opened the pull request:
Abstract DB wrapper with LevelDB implementation.
Tests included.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/NodeFactoryIo/ChainGuardian?pullRequest=12) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/NodeFactoryIo/ChainGuardian?pullRequest=12) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/NodeFactoryIo/ChainGuardian?pullRequest=12) it.</sub>

mpetrunic commented on package.json:
```diff
@@ -103,19 +103,31 @@
     "@babel/preset-env": "^7.4.5",
     "@babel/preset-react": "^7.0.0",
     "@babel/preset-typescript": "^7.3.3",
+    "@chainsafe/eth2.0-config": "^0.1.0",
```
should not be devDependency

mpetrunic commented on package.json:
```diff
@@ -103,19 +103,31 @@
     "@babel/preset-typescript": "^7.3.3",
+    "@chainsafe/eth2.0-config": "^0.1.0",
+    "@chainsafe/eth2.0-types": "^0.1.0",
+    "@chainsafe/ssz": "^0.5.1",
```
should not be devDependency

mpetrunic commented on package.json:
```diff
@@ -103,19 +103,31 @@
+    "chai-as-promised": "^7.1.1",
     "cross-env": "^5.1.3",
     "css-loader": "^2.1.1",
+    "deepmerge": "^4.0.0",
```
should not be devDependency

mpetrunic commented on src/db/api/abstract.ts:
```diff
@@ -0,0 +1,26 @@
+
+export abstract class DatabaseService implements IService {
+    protected config: IBeaconConfig;
+    protected db: IDatabaseController;
```
both properties could have readonly modifier

mpetrunic commented on src/db/controller/impl/level.ts:
```diff
@@ -0,0 +1,94 @@
+        super();
+        this.opts = opts;
+        this.logger = logger;
+        this.db = opts.db || level(opts.name || 'beaconchain', { keyEncoding: 'binary', valueEncoding: 'binary' });
```
don't think beaconchain is correct default name :)

mpetrunic commented on src/db/options.ts:
```diff
@@ -0,0 +1,27 @@
+    name: string;
+}
+
+export const DatabaseOptions: IConfigurationModule = {
```
We don't need this, this is for cli

mpetrunic commented on src/db/schema.ts:
```diff
@@ -0,0 +1,58 @@
+    merkleTree, // depositIndex -> MerkleTree
+    // validator
+    lastProposedBlock,
+    proposedAttestations
```
clear those buckets

mpetrunic commented on src/db/schema.ts:
```diff
@@ -0,0 +1,58 @@
+
+    finalizedBlock,
+    justifiedBlock,
+    progressiveMerkleTree
```
clear those keys

mpetrunic commented on src/logger/winston.ts:
```diff
@@ -0,0 +1,118 @@
+import { defaultLogLevel, LogLevel, ILogger, ILoggerOptions } from './interface';
+import chalk from 'chalk';
+
+export class WinstonLogger implements ILogger {
```
We don't wan't to use winston as we need something cross OS compatible (storing location and managing rotation on different OS) so don't implement logger interface just yet

mpetrunic commented on src/util/config.ts:
```diff
@@ -0,0 +1,78 @@
+export interface IConfigurationField<T = unknown> {
```
we don't need this file entirely, it's for CLI

mpetrunic commented on src/db/api/repository.ts:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    public async set(id: Id, value: T): Promise<void> {
+        await this.db.put(encodeKey(this.bucket, id), serialize(value, this.type));
```
could you extract serialize/deserialize/hashTreeRoot into separate class (something like "CGSerialization) so we can change serialization to google protobufs at some point.

mpetrunic commented:
@aostrun I think all those things should be under /src/renderer/services

aostrun commented:
Fixed all issues and implemented `CGSerialization<T>` interface. I tried creating a generic interface for serialization where type T represents a wrapper for all types supported by the serializer implementation. E.g. `SSZSerializer` has type T set to `AnySSZType`.
I'm not too happy with how this ends up looking inside the abstract `Repository` class as we have to set the serializer dependency `serializer: CGSerialization<any>`  generic type to `any` as we want to support various serializers.
@mpetrunic What do you think about this?

mpetrunic commented on src/renderer/services/db/abstract.ts:
```diff
@@ -0,0 +1,7 @@
+
+export interface CGSerialization<T> {
+    serialize(value: any, type: T): Buffer;
+    deserialize(value: Buffer, type: T): any;
```
I will answer you question here. We haven't really solved this one in ssz library as generics in typescript are kinda limited. So we can leave `any` here. 

I would only suggest for deserialize method to be
`deserialize<R>(value: Buffer, type: T): R`
as it's much easier to set type than doing `as Something` everywhere

mpetrunic replied:
Also you can hardcode AnySSZType as type in CGSerializer (you don't need generics) because if we change serialization we will write adapter to convert AnySSZType to required schema rather than writing type descriptors for every type

mpetrunic commented on src/renderer/services/db/controller/impl/level.ts:
```diff
@@ -0,0 +1,89 @@
+import { LevelUp } from 'levelup';
+import { IDatabaseController, ISearchOptions } from '../interface';
+import { EventEmitter } from 'events';
+// @ts-ignore
```
you still need ignore even after you added level types?

aostrun replied:
There are no types for `level` package, or at least I wasn't able to find them :/

mpetrunic replied:
its ok then. I will create PR for fix of e2e tests with esModuleInterop flag. It will include a lot of style guide changes, hopefully it won't break things much

mpetrunic commented on src/renderer/services/db/schema.ts:
```diff
@@ -0,0 +1,50 @@
+    exit, // hash -> VoluntaryExit
+    transfer, // hash -> Transfer
+    proposerSlashing, // hash -> ProposerSlashing
+    attesterSlashing // hash -> AttesterSlashing
```
We can remove all Bucket values and keep it empty for now, until we start storing stuff

mpetrunic commented on src/renderer/services/db/schema.ts:
```diff
@@ -0,0 +1,50 @@
+
+    latestState,
+    finalizedState,
+    justifiedState
```
We can remove all Key values and keep it empty for now, until we start storing stuff

mpetrunic commented on src/renderer/services/db/serializers/ssz.ts:
```diff
@@ -0,0 +1,18 @@
+
+const SSZ = new SSZSerializer();
+
+export default SSZ;
```
don't use export default. I will try to update eslint ro prefer named export

Reasoning:
https://humanwhocodes.com/blog/2019/01/stop-using-default-exports-javascript-module/

mpetrunic commented on src/renderer/services/db/schema.ts:
```diff
@@ -0,0 +1,32 @@
+
+// Buckets are separate database namespaces
+export enum Bucket {
+    test
```
nope, hardcode some string manually in test

mpetrunic commented on package.json:
```diff
@@ -134,6 +131,7 @@
     "file-loader": "^3.0.1",
     "fork-ts-checker-webpack-plugin": "^1.3.4",
     "html-webpack-plugin": "^3.2.0",
+    "husky": "^2.3.0",
```
we removed that
